crates/sncast/src/state.rs

```diff
@@ -0,0 +1,392 @@
+use crate::helpers::constants::STATE_FILE_VERSION;
+use crate::response::structs::{DeclareResponse, DeployResponse, InvokeResponse};
+use anyhow::{anyhow, Result};
+use camino::Utf8PathBuf;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+use std::fs;
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct ScriptTransactionsSchema {
+    pub version: u8,
+    pub transactions: Option<ScriptTransactionEntries>,
+}
+
+impl ScriptTransactionsSchema {
+    pub fn append_transaction_entries(&mut self, tx_entries: ScriptTransactionEntries) {
+        match self.transactions {
+            Some(ref mut existing_entries) => {
+                existing_entries
+                    .transactions
+                    .extend(tx_entries.transactions);
+            }
+            None => {
+                self.transactions = Some(tx_entries);
+            }
+        }
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ScriptTransactionEntries {
+    pub transactions: HashMap<String, ScriptTransactionEntry>,
+}
+
+impl ScriptTransactionEntries {
+    #[must_use]
+    pub fn get(&self, key: &str) -> Option<&ScriptTransactionEntry> {
+        self.transactions.get(key)
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ScriptTransactionEntry {
+    pub name: String,
+    pub output: ScriptTransactionOutput,
+    pub status: ScriptTransactionStatus,
+    pub timestamp: u32,
+    pub misc: Option<HashMap<String, Value>>,
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+#[serde(tag = "type")]
+pub enum ScriptTransactionOutput {
+    InvokeResponse(InvokeResponse),
+    DeclareResponse(DeclareResponse),
+    DeployResponse(DeployResponse),
+    ErrorResponse(ErrorResponse),
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ErrorResponse {
+    pub message: String,
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub enum ScriptTransactionStatus {
+    // script executed successfully, transaction accepted/succeeded
+    Success,
+    // script executed successfully, transaction rejected/reverted
+    Fail,
```

Comment:
Do we intend for users to read state file? If so it may be beneficial to add info about whether the transaction was reverted or rejected, since reverted transactions alter the state of the blockchain and cost money

Reply:
> Do we intend for users to read state file? 

Not directly probably, but it will be possible

> If so it may be beneficial to add info about whether the transaction was reverted or rejected, since reverted transactions alter the state of the blockchain and cost money

This information is intended to be held in `ScriptTransactionOutput` -> `ErrorResponse`. There is an example test showing this off `test_load_or_create_state_file_exists_with_txs`, I thought having this there would be a pretty sane choice as this is actually a transaction output. The `ScriptTransactionStatus` on the other hand was meant to be a simple, quick way for us to determine the tx status (without going into its output) - and wheter such tx should be replayed later on (or not).

Does that make sense?

Reply:
yep, makes sense